pkg/csi/service/fcd/controller_test.go

```diff
@@ -0,0 +1,644 @@
+	var cfg vcfg.Config
+	model := simulator.VPX()
+
+	if multiDc {
+		model.Datacenter = 2
+		model.Datastore = 1
+		model.Cluster = 1
+		model.Host = 0
+	}
+
+	err := model.Create()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	model.Service.TLS = tlsConfig
+	s := model.Service.NewServer()
+
+	// STS simulator
+	path, handler := sts.New(s.URL, vpx.Setting)
+	model.Service.ServeMux.Handle(path, handler)
+
+	// vAPI simulator
+	path, handler = vapi.New(s.URL, nil)
+	model.Service.ServeMux.Handle(path, handler)
+
+	// Lookup Service simulator
+	model.Service.RegisterSDK(lookup.New())
+
+	cfg.Global.InsecureFlag = insecureAllowed
+
+	cfg.Global.VCenterIP = s.URL.Hostname()
+	cfg.Global.VCenterPort = s.URL.Port()
+	cfg.Global.User = s.URL.User.Username()
+	cfg.Global.Password, _ = s.URL.User.Password()
+	// Configure region and zone categories
+	if multiDc {
+		cfg.Global.Datacenters = "DC0,DC1"
+	} else {
+		cfg.Global.Datacenters = vclib.TestDefaultDatacenter
+	}
+	cfg.VirtualCenter = make(map[string]*vcfg.VirtualCenterConfig)
+	cfg.VirtualCenter[s.URL.Hostname()] = &vcfg.VirtualCenterConfig{
+		User:         cfg.Global.User,
+		Password:     cfg.Global.Password,
+		VCenterPort:  cfg.Global.VCenterPort,
+		InsecureFlag: cfg.Global.InsecureFlag,
+		Datacenters:  cfg.Global.Datacenters,
+	}
+
+	// Configure region and zone categories
+	if multiDc {
+		cfg.Labels.Region = "k8s-region"
+		cfg.Labels.Zone = "k8s-zone"
+	}
+
+	return cfg, func() {
+		s.Close()
+		model.Remove()
+	}
+}
+
+func configFromEnvOrSim(multiDc bool) (vcfg.Config, func()) {
+	cfg, ok := vcfg.ConfigFromEnv()
+	if ok {
+		return cfg, func() {}
+	}
+	return configFromSim(multiDc)
+}
+
+type SearchIndex struct {
+	*simulator.SearchIndex
+	vm *simulator.VirtualMachine
+}
+
+func (s *SearchIndex) FindByDnsName(req *types.FindByDnsName) soap.HasFault {
+	res := &methods.FindByDnsNameBody{Res: new(types.FindByDnsNameResponse)}
+	if req.VmSearch && strings.EqualFold(req.DnsName, s.vm.Name) {
+		res.Res.Returnval = &s.vm.Self
+	}
+	return res
+}
+
+func TestCompleteControllerFlow(t *testing.T) {
+	config, cleanup := configFromEnvOrSim(false)
+	defer cleanup()
+
+	connMgr := cm.NewConnectionManager(&config, nil)
+	defer connMgr.Logout()
+
+	c := &controller{
+		cfg:     &config,
+		connMgr: connMgr,
+	}
+
+	//context
+	ctx := context.Background()
+
+	// Get a simulator VM
+	myVM := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine)
+	vmName := myVM.Name
+
+	// Get a simulator DC
+	//mydc := simulator.Map.Any("Datacenter").(*simulator.Datacenter)
+
+	// Get a simulator DS
+	myds := simulator.Map.Any("Datastore").(*simulator.Datastore)
+
+	err := connMgr.Connect(ctx, config.Global.VCenterIP)
+	if err != nil {
+		t.Errorf("Failed to Connect to vSphere: %s", err)
+	}
+
+	search := object.NewSearchIndex(connMgr.VsphereInstanceMap[config.Global.VCenterIP].Conn.Client)
+	si := simulator.Map.Get(search.Reference()).(*simulator.SearchIndex)
```

Comment:
You can use `simulator.Map.SearchIndex()` here instead.

Reply:
Will update